File: solutions/scala/zebra-puzzle/1/src/main/scala/ZebraPuzzle.scala
object ZebraPuzzle:
    case class Solution(waterDrinker: Nationality, zebraOwner: Nationality)

    def solve: Solution =
        Solution(
            waterDrinker = findNationalityFor(_.drink == Drink.Water),
            zebraOwner   = findNationalityFor(_.pet == Pet.Zebra)
        )

    private def findNationalityFor(predicate: Person => Boolean): Nationality =
        find(predicate).nationality

    private def find(predicate: Person => Boolean): Person =
        solution.find(predicate).get

    private lazy val solution: Iterable[Person] =
        val candidates: Array[Person] =
            PersonFactory
                .createAll()
                .filter(person => Constraints.HouseRules.forall(rule => Tester.test(person, rule)))
                .toArray

        PersonSetEnumerator
            .enumerate(candidates)
            .find(set => Constraints.NeighborRules.forall(rule => Tester.test(set, rule)))
            .get

File: solutions/java/zebra-puzzle/1/src/main/java/ZebraPuzzle.java
import java.util.Arrays;
import java.util.List;

class ZebraPuzzle {

    private final List<Person> solution;

    public ZebraPuzzle() {

        Person[] candidates = PersonFactory.createAll()
                                           .filter(person ->
                                                           Arrays.stream(Constraints.HOUSE_RULES.toArray(new HouseRule[0]))
                                                                 .allMatch(rule -> Tester.test(person, rule))
                                           )
                                           .toArray(Person[]::new);

        this.solution = PersonSetFinder.findDistinctPersonSets(candidates)
                                       .filter(set ->
                                                       Arrays.stream(Constraints.NEIGHBOR_RULES.toArray(new NeighborRule[0]))
                                                             .allMatch(rule -> Tester.test(set, rule))
                                       )
                                       .findFirst()
                                       .orElseThrow()
                                       .stream()
                                       .toList();
    }

    public String getWaterDrinker() {
        return find(person -> person.drink() == Drink.Water)
                .nationality()
                .name();
    }

    public String getZebraOwner() {
        return find(person -> person.pet() == Pet.Zebra)
                .nationality()
                .name();
    }

    private Person find(java.util.function.Predicate<Person> predicate) {
        return solution.stream()
                       .filter(predicate)
                       .findFirst()
                       .orElseThrow();
    }
}